torchgeo/samplers/single.py

```diff
@@ -262,60 +316,107 @@ def __init__(
         num_chips = 0
         self.hits = []
         areas = []
-        for hit in self.index.intersection(tuple(self.roi), objects=True):
-            bounds = BoundingBox(*hit.bounds)
-            if (
-                bounds.maxx - bounds.minx >= self.size[1]
-                and bounds.maxy - bounds.miny >= self.size[0]
-            ):
-                if bounds.area > 0:
-                    rows, cols = tile_to_chips(bounds, self.size)
-                    num_chips += rows * cols
-                else:
-                    num_chips += 1
-                self.hits.append(hit)
-                areas.append(bounds.area)
+        # use only geospatial bounds for filering in case return_as_ts is set to True
+        filter_indices = slice(0,4) if dataset.return_as_ts else slice(0,6)
+
+        if isinstance(self.roi, list):
+            for area in self.roi:
+                for hit in self.index.intersection(tuple(area), objects=True):
+                    # Filter out hits in the index that share the same extent
+                    if  hit.bounds[filter_indices] not in [ht.bounds[filter_indices] for ht in self.hits]:
+                        bounds = BoundingBox(*hit.bounds)
+                        if (
+                            bounds.maxx - bounds.minx >= self.size[1]
+                            and bounds.maxy - bounds.miny >= self.size[0]
+                        ):
+                            if bounds.area > 0:
+                                rows, cols = tile_to_chips(bounds, self.size)
+                                num_chips += rows * cols
+                            else:
+                                num_chips += 1
+                            self.hits.append(hit)
+                            areas.append(bounds.area)
+        else:
+            for hit in self.index.intersection(tuple(self.roi), objects=True):
+                # Filter out hits in the index that share the same extent
+                if  hit.bounds[filter_indices] not in [ht.bounds[filter_indices] for ht in self.hits]:
+                    bounds = BoundingBox(*hit.bounds)
+                    if (
+                        bounds.maxx - bounds.minx >= self.size[1]
+                        and bounds.maxy - bounds.miny >= self.size[0]
+                    ):
+                        if bounds.area > 0:
+                            rows, cols = tile_to_chips(bounds, self.size)
+                            num_chips += rows * cols
+                        else:
+                            num_chips += 1
+                        self.hits.append(hit)
+                        areas.append(bounds.area)
```

Comment:
This shares a lot of functionality with the code in the geosampler, perhaps that we can create a function for it?